### PR TITLE
Convert json-formlets from scalaz to cats

### PR DESCRIPTION
**Pull request by charlesfrayne1**
TODO: need to decide whether to pull in a cats property testing library or just delete those tests.

### CODE REVIEW

File: src/test/scala/gov/wicourts/json/formlet/package.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+trait ScalaCheckInstances {
+
+  implicit val genMonad: Monad[Gen] = new Monad[Gen] {
+
+    override def pure[A](a: A): Gen[A] = Gen.const(a)
+
+    def flatMap[A, B](fa: Gen[A])(f: A => Gen[B]): Gen[B] = fa.flatMap(f)
+
+    def tailRecM[A, B](a: A)(f: (A) ⇒ Gen[Either[A, B]]): Gen[B] =
```
Comment (cwinebr) [inline src/test/scala/gov/wicourts/json/formlet/package.scala]: Doesn't `Gen` provide `tailRecM`?  Also, parens.

File: src/test/scala/gov/wicourts/json/formlet/package.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      f(a).flatMap {
+        case Right(b) => pure(b)
+        case Left(aa) => tailRecM(aa)(f)
+      }
+  }
+
+  implicit val genApply: Apply[Gen] = new Apply[Gen] {
+    //override def pure[A](a: A): Gen[A] = Gen.const(a)
```
Comment (cwinebr) [inline src/test/scala/gov/wicourts/json/formlet/package.scala]: stray comment

File: src/test/scala/gov/wicourts/json/formlet/package.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    def flatMap[A, B](fa: Gen[A])(f: A => Gen[B]): Gen[B] = fa.flatMap(f)
+
+    def tailRecM[A, B](a: A)(f: (A) ⇒ Gen[Either[A, B]]): Gen[B] =
+      f(a).flatMap {
+        case Right(b) => pure(b)
+        case Left(aa) => tailRecM(aa)(f)
+      }
+  }
+
+  implicit val genApply: Apply[Gen] = new Apply[Gen] {
```
Comment (cwinebr) [inline src/test/scala/gov/wicourts/json/formlet/package.scala]: Do we need the separate `Apply` instance?

File: src/main/scala/gov/wicourts/json/formlet/ValidationErrors.scala [hunk 4]
```diff
@@ -104,19 +103,19 @@ object ValidationErrors {
...
         case ObjectErrors(l) => collapseAll(l)
         case ArrayErrors(l) => collapseAll(l)
-      }
+      },
     )
   }
 
   def string(name: String, error: String): ValidationErrors =
-    ObjectErrors(List((name, FieldErrors(NonEmptyList(error)))))
+    ObjectErrors(List((name, FieldErrors(NonEmptyList.of(error)))))
```
Comment (cwinebr) [inline src/main/scala/gov/wicourts/json/formlet/ValidationErrors.scala]: Not a big deal, but this should probably be `NonEmptyList.one`.

File: src/main/scala/gov/wicourts/json/formlet/JsonBuilder.scala [hunk 5]
```diff
@@ -42,16 +39,16 @@ class JsonObjectBuilder private [formlet] (
...
-    def append(f1: JsonObjectBuilder, f2: => JsonObjectBuilder): JsonObjectBuilder =
+    override def combine(f1: JsonObjectBuilder, f2: JsonObjectBuilder): JsonObjectBuilder =
       new JsonObjectBuilder(f1.items ++ f2.items)
   }
 
-  implicit val jsonObjectBuildEqual: Equal[JsonObjectBuilder] = Equal.equal((a1, a2) =>
-    a1.items === a2.items
-  )
+  implicit val jsonObjectBuildEq: Eq[JsonObjectBuilder] =
+    Eq.instance((a1, a2) => a1.items === a2.items)
```
Comment (cwinebr) [inline src/main/scala/gov/wicourts/json/formlet/JsonBuilder.scala]: Why not `Eq.by(_.items)` here, like above?

File: src/main/scala/gov/wicourts/json/formlet/Formlet.scala [hunk 6]
```diff
@@ -202,84 +225,90 @@ object Formlet extends FormletSyntax {
...
     M[_]: Monad,
     I,
     V: Monoid,
-    E: Semigroup
-  ]: Monad[Formlet[M, \/, I, V, E, ?]] =
-    new Monad[Formlet[M, \/, I, V, E, ?]] {
-      override def map[A, B](a: Formlet[M, \/, I, V, E, A])(f: A => B): Formlet[M, \/, I, V, E, B] =
+    E: Semigroup,
+  ]: Monad[Formlet[M, Either, I, V, E, ?]] =
+    new Monad[Formlet[M, Either, I, V, E, ?]] with StackSafeMonad[Formlet[M, Either, I, V, E, ?]] {
```
Comment (cwinebr) [inline src/main/scala/gov/wicourts/json/formlet/Formlet.scala]: Is it stack-safe, though?

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala [hunk 7]
```diff
@@ -2,55 +2,48 @@ package gov.wicourts.json.formlet
...
+    fromJson: Json => Either[NonEmptyList[String], A],
     name: String,
-    value: Option[A]
+    value: Option[A],
   ): FieldFormlet[M, Option[A]] =
     Formlet { c =>
-      val result = optionT[NonEmptyList[String] \/ ?](
-        c
-          .flatMap(_.downField(name))
+      val result = OptionT( //optionT[Either[NonEmptyList[String], *]](
```
Comment (cwinebr) [inline src/main/scala/gov/wicourts/json/formlet/Forms.scala]: stray comment

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala [hunk 8]
```diff
@@ -2,55 +2,48 @@ package gov.wicourts.json.formlet
...
-        j.right[NonEmptyList[String]].ensure(
-          NonEmptyList(s"Field $name must be a(n) $descr")
-        )(
-          matches
-        ) >>= (fromJson)
-      ).run.validation
+          j.asRight[NonEmptyList[String]]
+            .ensure(NonEmptyList.of(s"Field $name must be a(n) $descr"))(matches)
+            .flatMap(fromJson)
+            .map(_.some),
```
Comment (cwinebr) [inline src/main/scala/gov/wicourts/json/formlet/Forms.scala]: Why is this necessary?

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala [hunk 9]
```diff
@@ -127,105 +125,115 @@ object Forms {
...
-        .compose[Tuple2[JsonArrayBuilder, ?]]
-        .compose[Validation[ValidationErrors, ?]]
-      M.map(X.traverse(l.zipWithIndex) { case ((i, x), idx) =>
-        M.map(x.mapResult((r, v) => (
-          r.leftMap(o => ValidationErrors.arrayErrors(List((idx, o)))),
-          JsonArrayBuilder.item(v.toJson)
-        )).run(i))(_.swap)
-      })(_.swap)
+      type G[AA] = M[(JsonArrayBuilder, Validated[ValidationErrors, AA])]
+      implicit val applicativeG =
```
Comment (cwinebr) [inline src/main/scala/gov/wicourts/json/formlet/Forms.scala]: Can we just stick with `X` (and no implicit) here?

File: src/test/scala/gov/wicourts/json/formlet/FormletSpec.scala [hunk 10]
```diff
@@ -1,110 +1,29 @@
 package gov.wicourts.json.formlet
```
Comment (cwinebr) [inline src/test/scala/gov/wicourts/json/formlet/FormletSpec.scala:1]: I didn't review any of these *Spec.scala files.

**Review by cwinebr — CHANGES_REQUESTED** [review]

File: src/test/scala/gov/wicourts/json/formlet/package.scala (hunk 1 above)
Comment (charlesfrayne1) [reply src/test/scala/gov/wicourts/json/formlet/package.scala]: Not that I know of. Doesn't compile if I try to use it and doesn't show up in https://javadoc.io/doc/org.scalacheck/scalacheck_2.11/1.14.1/index.html#org.scalacheck.Gen Are you seeing it somewhere?  
Fixed the parentheses and the weird arrow

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala (hunk 8 above)
Comment (charlesfrayne1) [reply src/main/scala/gov/wicourts/json/formlet/Forms.scala]: If you figure it out, I'd love to know. They're defined almost identically, but this also came up in the ccap3 conversion, with a flatMapF that compiled in scalaz despite there being no Options around to use. Seems like it ought to have been broken from the start, since none of right, ensure or fromJson have any way of providing an Option to satisfy flatMapF

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala (hunk 9 above)
Comment (charlesfrayne1) [reply src/main/scala/gov/wicourts/json/formlet/Forms.scala]: I haven't found a way to get rid of G, but I can stick with X as the name to make the diff a little smaller.

File: src/test/scala/gov/wicourts/json/formlet/package.scala (hunk 1 above)
Comment (cwinebr) [reply src/test/scala/gov/wicourts/json/formlet/package.scala]: It is on the companion object.

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala (hunk 9 above)
Comment (cwinebr) [reply src/main/scala/gov/wicourts/json/formlet/Forms.scala]: Looks good.  (Am I missing a way to mark a MR comment as resolved?)

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala (hunk 7 above)
Comment (cwinebr) [reply src/main/scala/gov/wicourts/json/formlet/Forms.scala]: Looks good.

File: src/main/scala/gov/wicourts/json/formlet/JsonBuilder.scala (hunk 5 above)
Comment (cwinebr) [reply src/main/scala/gov/wicourts/json/formlet/JsonBuilder.scala]: Looks good.

File: src/main/scala/gov/wicourts/json/formlet/ValidationErrors.scala (hunk 4 above)
Comment (cwinebr) [reply src/main/scala/gov/wicourts/json/formlet/ValidationErrors.scala]: There are quite a few of these in this MR (if you want to throw sed at it).

File: src/test/scala/gov/wicourts/json/formlet/package.scala (hunk 3 above)
Comment (cwinebr) [reply src/test/scala/gov/wicourts/json/formlet/package.scala]: Looks good.

File: src/test/scala/gov/wicourts/json/formlet/package.scala (hunk 2 above)
Comment (cwinebr) [reply src/test/scala/gov/wicourts/json/formlet/package.scala]: No longer applicable.

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala (hunk 8 above)
Comment (cwinebr) [reply src/main/scala/gov/wicourts/json/formlet/Forms.scala]: Can we do away with `OptionT` and the `.map(_.some)`?

File: src/test/scala/gov/wicourts/json/formlet/FormletSpec.scala (hunk 10 above)
Comment (cwinebr) [reply src/test/scala/gov/wicourts/json/formlet/FormletSpec.scala:1]: I noticed there is some commented out code in them.

File: src/test/scala/gov/wicourts/json/formlet/package.scala (hunk 1 above)
Comment (charlesfrayne1) [reply src/test/scala/gov/wicourts/json/formlet/package.scala]: maybe you're looking at a different version? Or maybe I'm doing something wrong. I checked again and I still get  
value tailRecM is not a member of object org.scalacheck.Gen

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala (hunk 9 above)
Comment (charlesfrayne1) [reply src/main/scala/gov/wicourts/json/formlet/Forms.scala]: Not sure. I have a resolve button below each thread.

File: src/main/scala/gov/wicourts/json/formlet/ValidationErrors.scala (hunk 4 above)
Comment (charlesfrayne1) [reply src/main/scala/gov/wicourts/json/formlet/ValidationErrors.scala]: Definitely don't trust myself to do that and *save* effort. I'll get a scalafix recipe for it in the main project

File: src/test/scala/gov/wicourts/json/formlet/package.scala (hunk 1 above)
Comment (cwinebr) [reply src/test/scala/gov/wicourts/json/formlet/package.scala]: Perhaps it isn't available in the version we're using...?  It is definitely in the documentation you provided, though.

File: src/main/scala/gov/wicourts/json/formlet/Formlet.scala (hunk 6 above)
Comment (charlesfrayne1) [reply src/main/scala/gov/wicourts/json/formlet/Formlet.scala]: I think it was. But had enough doubts to rewrite it

File: src/main/scala/gov/wicourts/json/formlet/Forms.scala (hunk 8 above)
Comment (charlesfrayne1) [reply src/main/scala/gov/wicourts/json/formlet/Forms.scala]: done

Comment (cwinebr) [reply src/main/scala/gov/wicourts/json/formlet/Forms.scala]: Looks good.
